### ISSUE TITLE
North star: all tools should be able to run multiple times independently within a single JVM process

### ISSUE THREAD

**Issue by ahelwer** [tlaplus/tlaplus#946]
This is related to #891, which can be seen as a necessary prerequisite. As discussed in the last monthly community meeting:

It would be amazing to be able to run all the tools, especially TLC, several times within the same JVM process. Reasons:

1. Greatly improved development experience for other tools wishing to integrate these as a library
2. Opens door to interesting tests where you can compare TLC's output across multiple runs in a single test without having to use runtime class reloading nonsense
3. Much faster unit testing runs without the overhead of having to start up a new JVM process with every test

This should be considered a long term goal that is unlikely to be achieved within the next couple of years. However, it should be top of mind when adding new functionality or refactoring existing functionality.

**Comment by lemmy** [comment]
What are these other tools that wish to integrate TLC as a library?

**Comment by ahelwer** [comment]
Apalache does, I think. And with the difficulty of writing a TLA+ parser the door should be open for tools to do this. Having a language tooling ecosystem is a good thing. Otherwise why do we care about preserving the public API if nobody consumes it?

**Comment by ahelwer** [comment]
I recall some comment from you that we want to preserve public-facing APIs. However, I could be misremembering. I do not particularly care either way.

**Comment by lemmy** [comment]
Yes, Apalache is built with SANY, but I'm not aware that Apalache suffers from the problem/enhancement discussed in this PR>

**Comment by konnov** [comment]
> Yes, Apalache is built with SANY, but I'm not aware that Apalache suffers from the problem/enhancement discussed in this PR.

Apalache does not use TLC itself, though an ability to offload some computations to TLC could lead to something interesting, similar to concolic execution. The only thing that Apalache is using from `tla2tools.jar` is SANY. Indeed, we ran into filesystem-level race conditions when running SANY in multiple threads. I think the issue is currently solved by changed the temporary directory, or by having a mutex on top of SANY.

**Comment by lemmy** [comment]
The race condition mentioned by @konnov is tracked in https://github.com/tlaplus/tlaplus/issues/688, but is orthogonal to this is PR.  A related issue is https://github.com/tlaplus/tlaplus/issues/719.  The proposed removal of some SANY APIs is at https://github.com/tlaplus/tlaplus/issues/718.

**Comment by Calvin-L** [comment]
> What are these other tools that wish to integrate TLC as a library?

We've had a few users allude to such tools (or the desire to build such tools):
 - #756 ("To allow TLC to be building block in a larger system")
 - #814 ("my Recife project (that runs TLC)")
 - #941 ("I will try to create [a TLA+ reformatter]")

**Comment by lemmy** [comment]
Running TLC in the browser could also benefit from this enhancement. We should continue collecting scenarios and use cases to evaluate the potential impact. Achieving this will undoubtedly require significant effort.

**Comment by bugarela** [comment]
Add Quint to that list!

https://github.com/informalsystems/quint/issues/1424

**Comment by lemmy** [comment]
@bugarela https://github.com/informalsystems/quint/issues/1424 reads as if *multiple* invocations of TLC in the same VM is a secondary concern.  Primarily, you want an exchange format to translate a TLC counterexample (safety or liveness) to quint.  Do you agree?

**Comment by bugarela** [comment]
@lemmy I think there are two main concerns in that issue:
1. Invoking TLC from Quint CLI. We already do that for Apalache (through SHAI, the Apalache gRPC server) and keeping the same UI would be ideal.
2. Translating the TLC result into the Quint format.

Andrew's description here says:
> Greatly improved development experience for other tools wishing to integrate these as a library

and that's why I think this would be helpful for (1)

**Comment by lemmy** [comment]
Thanks for the clarification that multiple invocations in the same VM is a secondary concern.
